hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/BlobRenameHandler.java

```diff
@@ -0,0 +1,452 @@
+    abfsClient.deleteBlobPath(path, leaseId, tracingContext);
+    if (abfsLease != null) {
+      abfsLease.cancelTimer();
+    }
+    operatedBlobCount.incrementAndGet();
+    return true;
+  }
+
+  private void copyPath(final Path src, final Path dst, final String leaseId)
+      throws IOException {
+    String copyId;
+    try {
+      AbfsRestOperation copyPathOp = abfsClient.copyBlob(src, dst, leaseId,
+          tracingContext);
+      final String progress = copyPathOp.getResult()
+          .getResponseHeader(X_MS_COPY_STATUS);
+      if (COPY_STATUS_SUCCESS.equalsIgnoreCase(progress)) {
+        return;
+      }
+      copyId = copyPathOp.getResult()
+          .getResponseHeader(X_MS_COPY_ID);
+    } catch (AbfsRestOperationException ex) {
+      if (ex.getStatusCode() == HttpURLConnection.HTTP_CONFLICT) {
+        AbfsRestOperation dstPathStatus = abfsClient.getPathStatus(
+            dst.toUri().getPath(),
+            false, tracingContext, null);
+        if (dstPathStatus.getResult() != null && ((ROOT_PATH
+            + abfsClient.getFileSystem() + src.toUri()
+            .getPath()).equals(
+            getDstSource(dstPathStatus)))) {
+          return;
+        }
+      }
+      throw ex;
+    }
+    final long pollWait = abfsClient.getAbfsConfiguration()
+        .getBlobCopyProgressPollWaitMillis();
+    while (handleCopyInProgress(dst, tracingContext, copyId)
+        == BlobCopyProgress.PENDING) {
+      try {
+        Thread.sleep(pollWait);
+      } catch (Exception e) {
+
+      }
+    }
+  }
+
+  private String getDstSource(final AbfsRestOperation dstPathStatus) {
+    try {
+      String responseHeader = dstPathStatus.getResult()
+          .getResponseHeader(X_MS_COPY_SOURCE);
+      if (responseHeader == null) {
+        return null;
+      }
+      return new URL(responseHeader).toURI().getPath();
+    } catch (URISyntaxException | MalformedURLException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Verifies if the blob copy is success or a failure or still in progress.
+   *
+   * @param dstPath path of the destination for the copying
+   * @param tracingContext object of tracingContext used for the tracing of the
+   * server calls.
+   * @param copyId id returned by server on the copy server-call. This id gets
+   * attached to blob and is returned by GetBlobProperties API on the destination.
+   *
+   * @return true if copying is success, false if it is still in progress.
+   *
+   * @throws AzureBlobFileSystemException exception returned in making server call
+   * for GetBlobProperties on the path. It can be thrown if the copyStatus is failure
+   * or is aborted.
+   */
+  @VisibleForTesting
+  public BlobCopyProgress handleCopyInProgress(final Path dstPath,
+      final TracingContext tracingContext,
+      final String copyId) throws AzureBlobFileSystemException {
+    AbfsRestOperation op = abfsClient.getPathStatus(dstPath.toUri().getPath(),
```

Comment:
Generic: everywhere where getPathStatus is used, we can take care of whether implicit call is needed or not to reduce number of calls

Reply:
good point! taken.